internal/engine/k8swatch/pod_watch.go

```diff
@@ -37,79 +65,196 @@ type PodWatcher struct {
+type nsWatch struct {
+	watchers map[types.NamespacedName]bool
+	cancel   context.CancelFunc
 }
 
-type podWatchTaskList struct {
-	watcherTaskList
-	extraSelectors []ExtraSelector
+func selectorFromLabels(labelValues []v1alpha1.LabelValue) labels.Selector {
+	ls := make(labels.Set, len(labelValues))
+	for _, l := range labelValues {
+		ls[l.Label] = l.Value
+	}
+	return ls.AsSelector()
 }
 
-func (w *PodWatcher) diff(ctx context.Context, st store.RStore) podWatchTaskList {
-	state := st.RLockState()
-	defer st.RUnlockState()
-
-	w.mu.RLock()
-	defer w.mu.RUnlock()
+// OnChange reconciles based on changes to KubernetesDiscovery objects.
+func (w *PodWatcher) OnChange(ctx context.Context, st store.RStore, summary store.ChangeSummary) {
+	if len(summary.KubernetesDiscoveries.Changes) == 0 {
+		return
+	}
 
-	taskList := w.watcherKnownState.createTaskList(state)
+	w.mu.Lock()
+	defer w.mu.Unlock()
 
-	// TODO(nick): Fix PodWatcher to only watch in namespaces we've deployed to.
-	var extraSelectors []ExtraSelector
-	if len(taskList.watchableNamespaces) > 0 {
-		for _, mt := range state.Targets() {
-			for _, ls := range mt.Manifest.K8sTarget().ExtraPodSelectors {
-				if !ls.Empty() {
-					extraSelectors = append(extraSelectors, ExtraSelector{name: mt.Manifest.Name, labels: ls})
-				}
-			}
-		}
+	for key := range summary.KubernetesDiscoveries.Changes {
+		w.reconcile(ctx, st, key)
 	}
 
-	return podWatchTaskList{
-		watcherTaskList: taskList,
-		extraSelectors:  extraSelectors,
+	w.cleanupAbandonedNamespaces()
+}
+
+// HasNamespaceWatch returns true if the key is a watcher for the given namespace.
+func (w *PodWatcher) HasNamespaceWatch(key types.NamespacedName, ns k8s.Namespace) bool {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	nsWatch, ok := w.watchedNamespaces[ns]
+	if !ok {
+		return false
 	}
+	return nsWatch.watchers[key]
 }
 
-func (w *PodWatcher) OnChange(ctx context.Context, st store.RStore, _ store.ChangeSummary) {
-	taskList := w.diff(ctx, st)
+// ExtraSelectors returns the extra selectors for a given KubernetesDiscovery object.
+//
+// This is intended for use in tests.
+func (w *PodWatcher) ExtraSelectors(key types.NamespacedName) []labels.Selector {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	var ret []labels.Selector
+	for _, s := range w.extraSelectors[key] {
+		ret = append(ret, s.DeepCopySelector())
+	}
+	return ret
+}
 
+// Claimant returns the key that has uniquely claimed a particular K8s entity UID to prevent
+// duplicate events.
+//
+// This is intended for use in tests.
+func (w *PodWatcher) Claimant(uid types.UID) types.NamespacedName {
 	w.mu.Lock()
 	defer w.mu.Unlock()
+	return w.claimedUIDs[uid]
+}
+
+// reconcile manages namespace watches for the modified KubernetesDiscovery object.
+func (w *PodWatcher) reconcile(ctx context.Context, st store.RStore, key types.NamespacedName) {
+	state := st.RLockState()
+	kd := state.KubernetesDiscoveries[key]
+	st.RUnlockState()
+
+	if kd == nil {
```

Comment:
i liked the other pattern you used in FileWatch where you checked "if deleted || specChanged { tearDown() }"... i had trouble reasoning about the piecemeal diffing below

Reply:
Check out the latest version of the diff and let me know if you think it's easier to follow.

There's a temporary snag that I can't just fully rip down the old version because of the logic to only send `PodChangeEvent` if it's the first time it's being seen by the watcher (i.e. manifest/`KubernetesDiscovery`).

That'll go away imminently though once it swaps to `KubernetesDiscoveryUpdateStatusAction` since it'll just always emit a status update when there's a spec change. 🎉